### ISSUE TITLE
Improve Cassandra Example 

### ISSUE THREAD

**Issue by chrislovecnm** [kubernetes/kubernetes#34978]
Inside of K8s IP addresses are transient.  C\* expects fixed IP addresses.  I have an idea that needs to be tested. We may be able to use data on a PV and the C\* option `-Dcassandra.replace_address` when a pod is restarted, the C\* ring may adjust automatically.

**Comment by chrislovecnm** [comment]
@erictune do you want to assign this to me and label with C* example?

**Comment by mstump** [comment]
Spent some time thinking about this. Cassandra 2.1 introduced host IDs. With host IDs if a node comes up with a different IP than it had previously it'll automatically adjust and gossip out the new IP. We don't need to store the IP on the PV to guard against the lack of an IP guarantee in petsets.

However, in the instance that we lose the PV (ephemeral storage attached to a rack or machine) we'll still need to pass in the previous IP address or host ID so that the newly relocated pod can assume the identity of the dead node and stream data from the remaining replicas. This is still an unresolved issue.

**Comment by chrislovecnm** [comment]
What about an init container that adds annotations to the pod?  I am wondering if that is an option.  The pod is annotated with it's ip address.  You mentioned that you may be able to address this in C* ... that may be the best option.

**Comment by mstump** [comment]
I've got a working solution that uses the ready-probe.sh and run.sh scripts. I'll have a pull request ready tomorrow.

The ```ready-probe.sh``` script waits for C* to bootstrap then pulls the host ID. The script pushes that host ID to k8s as a label. 
```
  KUBE_TOKEN=$(</var/run/secrets/kubernetes.io/serviceaccount/token)
  HOST_ID=`cqlsh -e "select host_id from system.local;"|egrep '[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}'|awk '{$1=$1};1'`

LABEL_DATA=$(cat <<-END
  {
    "metadata":{
      "labels":{
        "cassandra-host-id":"$HOST_ID"
      }
    }
  }
END
)
  curl -k -v -XPATCH -H "Authorization: Bearer $KUBE_TOKEN" -H "Accept: application/json, */*" \
       -H "Content-Type: application/merge-patch+json" \
       --data "$LABEL_DATA" \
       https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_PORT_443_TCP_PORT/api/v1/namespaces/$POD_NAMESPACE/pods/$HOSTNAME
```

And then ```run.sh``` checks for the presence of the system tables on the data mount. If the node doesn't have data and the host ID is present as a label then we know that the node was previously bootstrapped but lost it's volume. If this condition is true the script will insert the host ID into the JVM args and the node will come up as a replacement and not a new node.

```
# if the data directory is empty and we know the host ID from the kubernetes
# lables then we know that we are a replacement node and not bootstraping a
# new node
if [ ! -d "$CASSANDRA_SYSTEM_PATH" ]; then
  CASSANDRA_HOST_ID=$(grep 'cassandra-host-id' "$KUBERNETES_LABELS_PATH" | egrep '[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}')
  if [ -n "$CASSANDRA_HOST_ID" ]; then
    $CASSANDRA_REPLACE_NODE="$CASSANDRA_HOST_ID"
  fi
fi
```

The downside to this approach is that if we lose the volume, and the pod is deleted we'll lose both the ID stored in the data files and the POD metadata. In this instance the node will be added as a new node and not a replacement.

The other option is a sidecar that monitors the cluster and stores this data in etcd. This is the approach taken in Netflix Priam, and the Mesos C* framework. The downside is that it requires more work.

I'm proposing that we take the script approach first, then add a sidecar at a later point.

**Comment by chrislovecnm** [comment]
We do have the ordinal index of the pods as well to rely on as well. Just thought about this ;) We have the size of the PetSet and the 0,1,2,3 ... Etc.

So a pod on start could know if it's neighbour to its right exists then it has already run once. 

Not sure if that data helps.  I would rather not run a side car, but use probes or init containers to get meta data around this more. Can we annotate the PetSet, rather than the pod?

**Comment by mstump** [comment]
Downside to that is that it ties us to deployment through petset. I'm working with a couple of installations that are pro-kubernetes but they're a little wary of petsets (too new). They want to use a replication controller, replication set or deployment and they've got high end shared storage.

Also, I think this General approach is applicable to ES and other databases. One thing that would be nice is to be able to put annotations on the parent of the pod in a way that can be generally abstracted across RC, RS, deployments, etc..

**Comment by chrislovecnm** [comment]
Petsets have been banged on pretty good, but yes they are new. The startup sequence is relay important, since you cannot boostrap two nodes at the same time.

**Comment by mstump** [comment]
I've got a working example where I'm using petset attributes to store Cassandra metadata. The branch is [here](https://github.com/mstump/kubernetes/commit/1c1a39b34cae487057248bb16ba042a36e968c38). It's not ready for commit because it breaks the non-petset examples and the manifest refers to my docker repo. I want to get sign off on the general idea before taking it to completion. 

This approach relies on a go [command line utility](https://github.com/mstump/cassandra-kubernetes-hostid) that finds the parent petset and modifies or reads the attributes. There are two consumers of the utility. The [ready-probe.sh](https://github.com/mstump/kubernetes/commit/1c1a39b34cae487057248bb16ba042a36e968c38#diff-3a1b0e1988f79997cb03549eb3f873ddR24) will use the utility to fetch the Cassandra host ID and then push that information as an attribute to the petset. 

The other consumer, [run.sh](https://github.com/mstump/kubernetes/commit/1c1a39b34cae487057248bb16ba042a36e968c38#diff-44770da8d7e937726613dd9f430b9db9R101) will check for the presence of Cassandra data, if it's not present it will then use the utility to check for a host ID for the node in the petset attributes. If the petset annotation is present then we know we're replacing a pre-existing host and we'll assume that node's identity.

**Comment by chrislovecnm** [comment]
I like the idea a lot.  A few thoughts.

- We need to figure out how to get the other examples to work, maybe a different container, or it checks to see if it is a pet?
- What are your thoughts on the CLI tool?  Should we move it into the k8s repo?
- We need more testing on both the CLI tool and in the Stateful set example in k8s.  Would be good to exercise this programmatically.  We are now depending on K8s functionality to run this Stateful Set, and if that functionality breaks, we need to have tests that break.

I am also checking on what is happening to the examples in the repo.  May want to put this into another repo.

**Comment by mstump** [comment]
I think we should probably put a stake in the ground and declare that StatefullSets are the preferred mechanism and deprecate the other examples. If that's not an option then I'd prefer to have a single image. It'll drive up complexity, but it'll be easier to maintain then 3-4 separate images. 

I've got mixed feelings about moving the code/image in or out of the repo. On one hand it's easier to iterate and grow the projects to include more functionality. For example, I don't think anybody has pushed the container to the repo after my last checkin and I can't update the example manifest. I'm willing to take on proactive development and ongoing maintenance if I can be moved up to member. I've got a vested interest in keeping up to date. I'll soon have production customers using the approach and will be allocating company resources to making it work. On the downside if we move it out then it's less likely that it'll be found, used or maintained. Even moving it to it's own repo within the k8s org has risks, there are lots of half-developed or abandoned projects. 

I'll look into the testing.

**Comment by chrislovecnm** [comment]
@zmerlynn my question is how we want to approach this.  The C* example is going to get some go code, and a new container.  Long term the plans are to move the examples out of core.  Do we have another issue open for this?  I would like to keep the image in gcr.io, because of the e2e testing.  We have cause a few Petset issues because of this example.  So here are my questions:

1. Do you want me to drop in a baby go project into the C* example
2. Do we want to look at pulling the example our of core, if so how do we do e2e

Thanks for your help as always!!

**Comment by chrislovecnm** [comment]
@zmerlynn any way that I can relabel issues like these?  We have C* labels :)

**Comment by cmluciano** [comment]
/close

closing since issue has not be active for a long time

**Comment by VinceMD** [comment]
We are trying to deploy Cassandra based on the example with PersistentVolume. Unfortunately if we kill a VM, and a new VM (EC2) gets created, PVC re-attached, Cassandra complains:

`java.lang.RuntimeException:` A node required to move the data consistently is down (/10.2.41.13).`

I think the new pod got a different IP and Cassandra is not happy. 
So this is a problem. I didn't find a way in Statefulset to set a static IP.

Is there another solution to this? Happy to share our improved Cassandra example if someone could give us pointers to go around that issue.

**Comment by VinceMD** [comment]
We run another test which consists in terminating a pod rather than terminating a VM/EC2. We obtained the same result: C* complaining about a node (understand here a C* cluster node) with a certain IP not being available. :(

**Comment by ipedrazas** [comment]
What we did is to use Calico and deployments where you can set a fixed IP. The problem is, as you say, that the pod dies and the new one has a new IP which for C* is a new (Cassandra) "node". There's nothing wrong with that approach, the problem comes when you want to bring back an old node with a different IP (to avoid data movement).

Running C* in k8s is possible but not straight forward. C* is not very happy in a very dynamic environment, so, you will have to make a few decisions about how to manage your cluster.

**Comment by VinceMD** [comment]
Anyone has an opinion on this approach? 
I altered 2 settings in run.sh:
`CASSANDRA_LISTEN_ADDRESS="${HOSTNAME}.${CASSANDRA_CLUSTER_NAME}.${POD_NAMESPACE}.svc.cluster.local"
CASSANDRA_BROADCAST_ADDRESS="${HOSTNAME}.${CASSANDRA_CLUSTER_NAME}.${POD_NAMESPACE}.svc.cluster.local"`

**Comment by jcharlytown** [comment]
I am unable to reproduce this issue when terminating a pod using `kubectl delete pod cassandra-1` (out of 0,1,2). We run on Azure ACS and make use of PVCs, which get reattached after the pod has been recreated by the statefulset. The container image is `gcr.io/google-samples/cassandra:v12` and I have made sure that the IP address in fact changes when the pod is recreated.

Logs of recreated pod:
```
INFO  08:59:28 Starting up server gossip
INFO  08:59:29 Updating topology for /10.244.3.13
INFO  08:59:29 Updating topology for /10.244.3.13
INFO  08:59:29 Using saved tokens [-1769721479951149921, -2310477068061707012, -2611658706965722681, -2983853379763058460, -3654472369005609328, -3837218999163995936, -4555044558535339656, -4992177074760103974, -5194157912918549698, -6286170896675845177, -6448295961328602759, -7310746583630386025, -7547089987527487804, -8181605939416214551, -8656783814350920730, -8718395193707488594, 1575499785299607070, 3173340802516907645, 3248746770687953051, 3559635923893510446, 363803942318230525, 3960008722450945105, 4617873020243149163, 4726450396815483489, 4739810261617260529, 5165369818734516510, 5636709232502593665, 587118985700116648, 6011373453842679740, 6705467816625001903, 80554791961482909, 8281275030343513090]
INFO  08:59:29 Node /10.244.3.13 state jump to NORMAL
INFO  08:59:29 Waiting for gossip to settle before accepting client requests...
WARN  08:59:30 Not marking nodes down due to local pause of 12698447716 > 5000000000
INFO  08:59:30 Node /10.244.2.32 has restarted, now UP
INFO  08:59:30 Node /10.244.2.32 state jump to NORMAL
INFO  08:59:30 InetAddress /10.244.2.32 is now UP
INFO  08:59:30 Updating topology for /10.244.2.32
INFO  08:59:30 Updating topology for /10.244.2.32
INFO  08:59:30 Node /10.244.0.127 has restarted, now UP
INFO  08:59:30 Node /10.244.0.127 state jump to NORMAL
INFO  08:59:30 InetAddress /10.244.0.127 is now UP
INFO  08:59:30 Updating topology for /10.244.0.127
INFO  08:59:30 Updating topology for /10.244.0.127
INFO  08:59:30 Handshaking version with cassandra-0.cassandra.default.svc.cluster.local/10.244.0.127
INFO  08:59:30 Node /10.244.3.11 is now part of the cluster
WARN  08:59:30 Not updating host ID b1eadbb4-4269-4494-ab41-e5e080dc6073 for /10.244.3.11 because it's mine
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -1769721479951149921.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -2310477068061707012.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -2611658706965722681.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -2983853379763058460.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -3654472369005609328.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -3837218999163995936.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -4555044558535339656.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -4992177074760103974.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -5194157912918549698.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -6286170896675845177.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -6448295961328602759.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -7310746583630386025.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -7547089987527487804.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -8181605939416214551.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -8656783814350920730.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token -8718395193707488594.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 1575499785299607070.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 3173340802516907645.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 3248746770687953051.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 3559635923893510446.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 363803942318230525.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 3960008722450945105.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 4617873020243149163.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 4726450396815483489.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 4739810261617260529.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 5165369818734516510.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 5636709232502593665.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 587118985700116648.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 6011373453842679740.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 6705467816625001903.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 80554791961482909.  Ignoring /10.244.3.11
INFO  08:59:30 Nodes /10.244.3.11 and /10.244.3.13 have the same token 8281275030343513090.  Ignoring /10.244.3.11
INFO  08:59:31 InetAddress /10.244.3.11 is now DOWN
INFO  08:59:31 Handshaking version with /10.244.2.32
```

Logs of some other pod:
```
WARN  08:56:22 Unlogged batch covering 17 partitions detected against table [test.song]. You should use a logged batch for atomicity, or asynchronous writes for performance.
INFO  08:58:58 InetAddress /10.244.3.11 is now DOWN
INFO  08:58:59 Handshaking version with /10.244.3.11
INFO  08:59:27 Handshaking version with /10.244.3.13
INFO  08:59:27 Handshaking version with /10.244.3.13
INFO  08:59:30 Node /10.244.3.13 is now part of the cluster
WARN  08:59:30 Host ID collision for b1eadbb4-4269-4494-ab41-e5e080dc6073 between /10.244.3.11 and /10.244.3.13; /10.244.3.13 is the new owner
INFO  08:59:30 InetAddress /10.244.3.13 is now UP
INFO  08:59:30 Updating topology for /10.244.3.13
INFO  08:59:30 Updating topology for /10.244.3.13
INFO  09:00:31 FatClient /10.244.3.11 has been silent for 30000ms, removing from gossip
```

Is there anything I am not getting? If so, it would be important to know since at some point we're going to loose a VM and I'd rather be prepared :)

**Comment by dcparker88** [comment]
@VinceMD I'm starting down this path as well, creating a 3 node c* cluster in kubernetes, and running in to a lot of the same issues (pods won't rejoin the cluster after being deleted, etc.) - are you having any luck with the changes you made? I'm looking to automate as much of this as I can, since we will be deploying many clusters, and running manual administrative tasks on them isn't a good option. 

So a few questions if you have time:
* are you using StatefulSets?
* are you still using a custom `run.sh` and `ready-probe.sh`?

**Comment by VinceMD** [comment]
@dcparker88 I'm still using statefulsets. I only changed  CASSANDRA_LISTEN_ADDRESS and CASSANDRA_BROADCAST_ADDRESS as per my comment above. So far so good but we are just starting load testing and disaster recovery (self healing after loosing nodes or pods) testing.
Also need to load bunch of data to see how that behaves.

**Comment by VinceMD** [comment]
@dcparker88 I guess my fix does not reduce data movement. 

I'm starting to spend more time on Cassandra.

is there an update on 

"Do we want to look at pulling the example our of core, if so how do we do e2e"

And is there a slack channel when we can collaborate?

I have an interest to see the project grow and scale in Kubernetes StatefulSet.

Anyone still thinks StatefulSet is not great for C* and has a traditional deployment/replica recipe to share?

**Comment by dcparker88** [comment]
@VinceMD I'm starting more and more testing as well - and I am not using a StatefulSet. Part of that reason is I am using Spinnaker to deploy, and ours doesn't have support for SS yet. I'm basically creating my own SS:

* I'm creating 3 different deployments, with 1 pod each (for a total of 3 pods)
* Each of these has a hostpath volume that is statically defined (/data/node1, /data/node2, etc)
* these hostpaths all go back to the same NAS - so regardless of which k8s host my pod lives on, the data drive is the same
  * this way, if the pod dies, or anything happens, it comes back with the same data directory
  * cassandra 3.x is much better at handling IP address changes if it starts up with the same data directory, and you don't need to do replace.address 

This is working well so far from a stability perspective. I no longer have to removenode when a pod dies. This does have some issues still though:
* If I want to deploy a change/upgrade/etc I either need to rebuild the whole cluster, or deploy new pods on new data directories 
* I lose the ability to dynamically scale the replicaset

I'm still experimenting with this and know what I have is far from perfect, but it's working for me so far.
